### ISSUE TITLE
Scores based masking in generator is broken

### ISSUE THREAD

**Issue by inspirit** [lucidrains/soundstorm-pytorch#40]
I think the scores based mask adjusting is broken now:
https://github.com/lucidrains/soundstorm-pytorch/blob/1616cdfa50ef80ae45f5984a6d70af8b55efcc24/soundstorm_pytorch/soundstorm.py#L957-L967

`scores` will always be fully filled with `mask_value` and so any further sorting and mask adjusting will not do anything meaningful. This happens because `mask` is set zeros first and then within this block:
```
if not self.can_mask_prev_unmasked:
    scores = scores.masked_fill(~mask, mask_value)
```
we reset all `scores` to `mask_value`

as a result we never have any scores except `mask_value`

**Comment by lucidrains** [comment]
@inspirit thank you for this

actually a pretty big bug and i wonder how Steven and Lucas even got results with this hanging around

could you check to see if `prev_mask` addresses this? i've also turned it off, as i don't think the paper had this anyways

**Comment by lucidrains** [comment]
@inspirit thank you for this

actually a pretty big bug and i wonder how Steven and Lucas even got results with this hanging around

could you check to see if `prev_mask` addresses the issue? i've also turned it off as it isn't essential

**Comment by lucidrains** [comment]
@inspirit while you are looking at this, could you run the two side by side and let us know which one performs better? i may even remove it if it does not do anything

**Comment by inspirit** [comment]
yeah should be fine now, i was wondering how all the people reported good results so far :)

okay, let me try running 2 experiments, but honestly I dont think it will make much of a difference

**Comment by lucidrains** [comment]
> yeah should be fine now, i was wondering how all the people reported good results so far :)
> 
> okay, let me try running 2 experiments, but honestly I dont think it will make much of a difference

Lucas must have had the flag turned on, unless if that bug was introduced in some commit that came after his experiments

Steven works for some startup, and in that scenario, sometimes researchers / companies quietly fix things without submitting anything upstream

**Comment by inspirit** [comment]
i will need to train few epochs first to have it generate anything good :)

**Comment by lucidrains** [comment]
@inspirit how did it go?

**Comment by lucidrains** [comment]
thanks! 👌
